### ISSUE TITLE
Переписать механизм конфигурирования движка и лоадер

### ISSUE THREAD

**Issue by KarelWintersky** [everypony/tabun#153]
Конфиг должен храниться в понятной совокупности файлов.

Лоадер должен создавать структуру конфига на основе этой совокупности файлов и задавать вычислимые значения на основе этих данных.

Композер загружается отдельно.

**Comment by KarelWintersky** [comment]
Пример:
```
$config['locale']['timezone'] = 'Europe/Moscow';
```
Содержит таймзону сервера.

Где-то в коде есть такой хардкод:
```
'date' => (string) $oVote->getDate().'+03:00',
```

Хардкод это плохо. Не вмешиваясь в потроха ливстрита, лучше переписать это как-то так:
```
'date' => (string) $oVote->getDate() . Config::Get('locale.timezonestring'),
```

Значение конфигурационной переменной `locale.timezonestring` определяется в лоадере.

P.S. Сразу подчеркну, это костыль для костыльного ливстрита.

**Comment by andreymal** [comment]
Причём оба варианта в примере неверные, потому что до 2014 года было очень даже +04:00 )

Должно быть что-то вроде `date('c', strtotime($oVote->getDate()))`

@Farxial это тебе к сведению наверно

**Comment by Farxial** [comment]
Метка времени голоса хранится [в формате DATETIME](https://github.com/everypony/tabun/blob/development/fixtures/migrations/0000_schema.sql#L1183), т.е. [без указания часового пояса и в локальном времени](https://dev.mysql.com/doc/refman/8.0/en/datetime.html):
> MySQL converts TIMESTAMP values from the current time zone to UTC for storage, and back from UTC to the current time zone for retrieval. (This does not occur for other types such as DATETIME.)

*Если* когда-то на сервере неправильно осуществлялась запись времени, а информации о часовом поясе нет, разве есть шанс исправить это сейчас?

В текущей реализации клиенту передаётся временная зона +03:00, _потому что предполагается, что она (всегда) использовалась при записи в БД_, [на клиенте она парсится](https://github.com/everypony/tabun/blob/development/frontend/app/vote.coffee#L120), клиент обрабатывает её и отображает в своём часовом поясе, соответствующем настройкам пользователя ОС и/или браузера. 

... Может, тогда стоит заменить обработку метки времени на клиенте на tpl-шаблон (только для метки времени или для всего списка голосов), где сам Smarty поставит, например, `{date_format date=$oCommentDate format="j F Y, H:i:s"}`?
Это закроет данный вопрос, хотя от рассогласования меток времени в случае, если оно имеет место, не спасёт.

**Comment by andreymal** [comment]
@Farxial локальное время — московское, сведения о записи отличного от московского времени мне не известны

>предполагается, что она (всегда) использовалась при записи в БД

Неправильно предполагается, повторю — до 2014-го московское время было +04:00

>Smarty поставит метку

Я свой пример своровал как раз из исходников Smarty)

**Comment by Farxial** [comment]
> Неправильно предполагается, повторю — до 2014-го московское время было +04:00

А что делать? ![https://files.everypony.ru/smiles/f4/b4/1a9be0.png](https://files.everypony.ru/smiles/f4/b4/1a9be0.png)
Неправилен подход, когда метки времени жёстко привязываются к значению временной зоны, а само значение меняется - переменная используется в качестве константы.
Тогда, может, если рассогласование имеет место, их должен устранить админ, заменив часть меток времени или вовсе переведя все метки времени в зоно-независимый формат?

**Comment by andreymal** [comment]
>а само значение меняется

Значение не менялось — и до 2014-го, и после было московское время. Рассогласований на Табуне нет (по крайней мере инфы об обратном у меня нет)

>или вовсе переведя все метки времени в зоно-независимый формат?

Это, конечно, самое правильное, но надо кому-то прод очень аккуратно конвертировать, накосячить в октябре 2014-го очень легко

**Comment by Farxial** [comment]
> Значение не менялось — и до 2014-го, и после было московское время.

Значение часового пояса - его смещение от некоторого абсолютного положения, одинакового всегда и для всех.
Чтобы не получилось так, что метка времени некоторого события, спустя несколько лет, отображалась увеличенной или уменьшенной, пусть даже на час, при чём у всех пользователей, включая даже не испытывающих проблем с часовыми поясами, нужно либо не трогать часовой пояс, либо просто не использовать.

> Рассогласований на Табуне нет (по крайней мере инфы об обратном у меня нет)

Тогда, может, и "+03:00" - норм?)
(Хотя вопрос о переводе на tpl всё равно пока открыт.)

> Это, конечно, самое правильное, но надо кому-то прод очень аккуратно конвертировать, накосячить в октябре 2014-го очень легко

Не спорю...

**Comment by andreymal** [comment]
>Значение часового пояса - его смещение

Нет, значение часового пояса — это регион (кроме UTC и парочки других). В случае Табуна это не +03:00 и не +04:00, а очень даже Europe/Moscow. Уже по этому региону, дате (до/после 2014-го) и наличию/отсутствию летнего времени вычисляется конкретное смещение. Попробуй уже запусти с разными датами код, который я показал ;)

>"+03:00" - норм?)

Мне опять написать, что было до 2014-го?)

---

А вообще давай с оффтопиком завязывать, я так, мимоходом упомянул, раз уж этот код подняли. Чинить его надо не в рамках конфига с лоадером

**Comment by Farxial** [comment]
> Попробуй уже запусти с разными датами код, который я показал ;)

Выводит разные временные зоны. Да, тогда, наверное, действительно может помочь.
![https://files.everypony.ru/smiles/42/5e/820371.png](https://files.everypony.ru/smiles/42/5e/820371.png)
Спасибо ^^

> А вообще давай с оффтопиком завязывать

Я не против. Если честно, мне вообще не оч удобно спорить из браузера в виртуалке без русской локали, проброшенного через SSH и Xming (но зато "нового" и поддерживаемого GitHub-ом) ![https://files.everypony.ru/smiles/42/5e/820371.png](https://files.everypony.ru/smiles/42/5e/820371.png)
